packages/financial-templates-lib/src/price-feed/DefiPulseTotalPriceFeed.js

```diff
@@ -0,0 +1,145 @@
+      let past = this.historicalPrices[i].timestamp;
+      let val = this.historicalPrices[i].tvlUSD;
+
+      if (past > closestTime.timestamp && past < time) {
+        closestTime.timestamp = past;
+        closestTime.tvlUSD = val;
+      }
+    }
+
+    const historicalPrice = this.scaleResult(closestTime.tvlUSD);
+
+    if (closestTime.timestamp === 0) {
+      return undefined;
+    } else {
+      return historicalPrice;
+    }
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  async update() {
+    const currentTime = this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+      message: "Updating",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // 1. Construct URLs.
+    const priceUrl = `https://data-api.defipulse.com/api/v1/defipulse/api/GetHistory?period=1w&api-key=${this.apiKey}`;
+
+    // 2. Send requests.
+    const response = await this.networker.getJson(priceUrl);
+
+    // 3. Check responses.
+    if (!response) {
+      throw new Error(`🚨Could not parse price result from url ${priceUrl}: ${JSON.stringify(response)}`);
+    }
+
+    // 4. Parse results.
+    // Return data structure:
+    //        [{
+    //            "timestamp":"1611511200"
+    //            "tvlUSD":25583565042,
+    //             ...
+    //          },
+    //           {
+    //            "timestamp":"1611507600",
+    //            "tvlUSD":25177860561
+    //             ...
+    //          },
+    //        ]
+
+    // Get tvlUSD for most most recent timestamp
+    let mostRecent = { timestamp: 0, tvlUSD: 0 };
+    for (let i = 0; i < response.length; i++) {
+      if (Number(response[i].timestamp) > mostRecent.timestamp) {
+        mostRecent.timestamp = Number(response[i].timestamp);
+        mostRecent.tvlUSD = Number(response[i].tvlUSD);
+      }
+    }
+
+    const newPrice = this.scaleResult(mostRecent.tvlUSD);
+
+    // 5. Store results.
+    this.lastUpdateTime = currentTime;
+    this.currentPrice = newPrice;
+    this.historicalPrices = response;
+  }
+
+  scaleResult(_tvlUSD) {
+    // As described in UMIP 24
+    // In an effort to make the token price affordable, the value of the token is the tvlUSD divided by 1 billion.
+    // We also cut off precision after 3 decimals to match the specified price step of .001
+
+    return this.toWei((_tvlUSD / 1000000000).toFixed(3), "ether");
```

Comment:
nit for readability:
```suggestion
    const billion = 1_000_000_000;
    return this.toWei((_tvlUSD / billion).toFixed(3), "ether");
```

Reply:
Changed.

Reply:
Do most synths like this use USDC or TUSD? I had originally thought of using USDC (6 decimals).

Reply:
Any tips for signing off on old commits? Just set up a GPG keypair. Should be good signing off from here on.